tests/Validation/ValidationTest.php

```diff
@@ -0,0 +1,196 @@
+
+$frameworksDir = realpath(__DIR__ . '/../../validation/frameworks');
+
+class ValidationTest extends TestCase
+{
+    public function frameworkErrorProvider()
+    {
+        global $frameworksDir;
+        $frameworks = glob($frameworksDir . '/*', GLOB_ONLYDIR);
+
+        $testProviderArray = array();
+        foreach ($frameworks as $frameworkDir) {
+            $frameworkName = basename($frameworkDir);
+            if ($frameworkName !== '_cases') {
+                continue;
+            }
+
+            $iterator = new RecursiveDirectoryIterator($frameworkDir);
+            $skipped = json_decode(file_get_contents(__DIR__ . '/skipped.json'));
+
+            foreach (new RecursiveIteratorIterator($iterator) as $file) {
+                if (strpos(\strrev((string)$file), \strrev(".php")) === 0 && !\in_array(basename((string)$file), $skipped)) {
+                    if ($file->getSize() < 100000) {
+                        $testProviderArray[$frameworkName . "::" . $file->getBasename()] = [$file->getPathname(), $frameworkName];
+                    }
+                }
+            }
+        }
+
+        if (count($testProviderArray) === 0) {
+            throw new Exception("ERROR: Validation testsuite frameworks not found - run `git submodule update --init --recursive` to download.");
+        }
+
+        return $testProviderArray;
+    }
+
+    /**
+     * @group validation
+     * @dataProvider frameworkErrorProvider
+     * @param $testCaseFile
+     * @param $frameworkName
+     */
+    public function testDefinitionErrors($testCaseFile, $frameworkName)
+    {
+        $fileContents = file_get_contents($testCaseFile);
+        $actualValues = $this->getActualTestValues($testCaseFile, $fileContents);
+
+        $outputFile = getExpectedValuesFile($testCaseFile);
+        if (!file_exists($outputFile)) {
+            file_put_contents($outputFile, json_encode($actualValues, JSON_PRETTY_PRINT|JSON_UNESCAPED_SLASHES));
+        }
+
+        $expectedValues = (array)json_decode(file_get_contents($outputFile));
+
+        try {
+            $this->assertEquals($expectedValues['definitions'], $actualValues['definitions']);
+
+            try {
+                $this->assertArraySubset((array)$expectedValues['references'], (array)$actualValues['references'], false, 'references don\'t match.');
+            } catch (\Throwable $e) {
+                $this->assertEquals((array)$expectedValues['references'], (array)$actualValues['references'], 'references don\'t match.');
+            }
+        } catch (\Throwable $e) {
+            $outputFile = getExpectedValuesFile($testCaseFile);
+            if ($frameworkName === '_cases') {
+                file_put_contents($outputFile, json_encode($actualValues, JSON_PRETTY_PRINT|JSON_UNESCAPED_SLASHES));
+            }
+
+            throw $e;
+        }
+    }
+
+    private function getActualTestValues($filename, $fileContents): array
+    {
+        global $frameworksDir;
```

Comment:
No globals please. I only see it being used in this class, so a private property should be fine.